### ISSUE TITLE
Fast path for device_put

### ISSUE THREAD

**Issue by jjyyxx** [google/jax#16905]
## Some background
I've been using JAX for several years to accelerate computation of various scale. For large transformer models, for my use case, the improvement is around 1.5x to 2x compared to PyTorch (thus not quite obvious). For smaller but complex scientific models, this often leads to orders of magnitude improvement.

For example, a recent workload runs on CPU and contains a lot of small computations. The jitted JAX function avoids the huge dispatch cost that PyTorch incurs. The workload also requires auto differentiation, thus Numba or Cython is not an option. In short, I was able to speed up the workload by 1000 times, from 20ms to 30us.

## Problem
Then I discovered that while the computation is fast, there was non-trivial overhead at JIT boundary. I used to prefer explicit conversion between `np.ndarray` and `jax.Array` at JIT boundary with `jax.device_put/jax.device_get` or `jnp.asarray/np.asarray`. But for my use case, the time cost of explicit conversion is even larger than the actual execution time!

Some quick benchmark below (`batch_context` is a PyTree with 8 arrays)
```python
# Method I used to prefer
%timeit jax.device_put(batch_context)
# 134 µs ± 657 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

# An alternative
%timeit jax.tree_util.tree_map(lambda x: jnp.asarray(x), batch_context)
# 345 µs ± 1.41 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```
A comparison with PyTorch
```python
# Much faster even copied on both Numpy and PyTorch side
%timeit jax.tree_util.tree_map(lambda x: torch.from_numpy(x.copy()).clone(), batch_context)
# 16.4 µs ± 21.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```
A potential fast path using some internals is much faster and comparable to PyTorch performance
```python
from jax._src.lib import xla_client as xc
from jax.core import ShapedArray
from jax.sharding import SingleDeviceSharding

device = jax.devices()[0]  # CpuDevice(id=0)
sharding = SingleDeviceSharding(device)

def fast_numpy_to_device_array(x):
    aval = ShapedArray(x.shape, x.dtype)
    return xc.batched_device_put(aval, sharding, [x], [device], True)

%timeit jax.tree_util.tree_map(fast_numpy_to_device_array, batch_context)
# 25.3 µs ± 35.8 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```
In addition, I found that directly passing PyTree with raw Numpy array leaves to the jitted function also works as expected, and only leads to negligible slowdown of serveral microseconds, which becomes the fastest way to invoke the jitted function.

## Goal
JAX gives me the impression that -- while the composition of functional transform is elegant and the jitted execution is efficient, something like JIT dispatch is less optimally handled. The previously landing CPP dispatch path has improved the dispatch efficiency a lot, but for array conversion, there may still be room for improvement. Or at least, it could be confirmed and documented somewhere that for simple unsharded usecase, passing the Numpy array directly is faster.

Please:

- [x] Check for duplicate requests.
- [x] Describe your goal, and if possible provide a code snippet with a motivating example.

**Comment by hawkinsp** [comment]
Yes, we're aware `device_put` is slower than it should be and looking at speeding it up soon.

One possible workaround for now: try `jit(lambda x: x)` or `jit(lambda x: x+0)`. The latter at least will force us to use the (probably faster) `jit` dispatch path.

**Comment by jjyyxx** [comment]
Yes, I was thinking of "abusing" jit for this purpose. But I only tried `jit(lambda x: x)`, and it turned out to be even slower! Thus I did not include that solution above. I did not realize that some "clever" optimization occurred (maybe by comparing Tracer identity?), and that somehow slowed down the execution. The extra `+ 0` or copy did the trick, making it the fastest explicit conversion solution (somewhat amusing).

The extra benchmark:
```python
f1 = jax.jit(lambda x: x); f1(batch_context)
f2 = jax.jit(lambda x: jax.tree_util.tree_map(jnp.copy, x)); f2(batch_context)
%timeit f1(batch_context)
# 234 µs ± 2.55 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
%timeit f2(batch_context)
# 12.3 µs ± 239 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

**Comment by hawkinsp** [comment]
Yes. If the `jaxpr` of the `jit` is trivial (has no equations), then we hit a "fast" path. But as you note, the "fast" path is no longer as fast as the path that actually dispatches a computation. We put in much more work optimizing the `jit` dispatch path to date and it shows!

**Comment by jon-chuang** [comment]
Actually, a potential root cause is identified here: https://github.com/openxla/xla/issues/12452#issuecomment-2110733606

I find that `batched_device_put` is much faster because in practice, in many systems, especially HPC/cloud, memcpy is actually slower than h2d transfer, and `device_put` or lazily getting pjrt to dispatch a h2d upon kernel launch incurs a memcpy.

I believe that more should be done to make the default behaviour zero-copy or else promote the use of `batched_device_put` over `device_put`.

**Comment by jon-chuang** [comment]
> The latter at least will force us to use the (probably faster) jit dispatch path.

For my purposes, jit dispatch path is just as slow as `device_put`

**Comment by lengstrom** [comment]
Hi, I am experiencing similar issues - is there a workaround for faster data transfer across devices? For context, I get roughly ~2.7GBps transfer speed with `device_put`; in contrast, writing a very simple CUDA program directly I get ~25GBps. Sample code for both approaches below. My hardware is an H100 (PCIe).

#### JAX:

```python
import time
import jax
import jax.numpy as jnp

cpu_dev = jax.devices('cpu')[0]

GBs = 10
def big_tensor():
    return jnp.ones((GBs * 1024**3 // 4,), dtype=jnp.float32, device=jax.devices('gpu')[0])

def test_transfer(state):
    jax.block_until_ready(state)
    s = time.time()

    out = jax.device_put(state, cpu_dev)

    jax.block_until_ready(out)
    dur = time.time() - s
    print(f"Time taken: {dur}; gbps: {GBs/dur}")

for i in range(5):
    test_transfer(big_tensor())
```
with output
```
Time taken: 3.7397873401641846; gbps: 2.6739488346310565
Time taken: 3.6008787155151367; gbps: 2.7770999220031807
Time taken: 3.677137613296509; gbps: 2.7195065976970936
Time taken: 3.594850778579712; gbps: 2.781756633567665
Time taken: 3.5868709087371826; gbps: 2.787945330187717
```

#### CUDA program

```c
#include <chrono>
#include <cstdio>
#include <cuda_runtime.h>

const uint64_t num_bytes = 10ul * (1ul << 30);  // 10 GB
const uint64_t num_floats = num_bytes / sizeof(float);

int main() {
    cudaStream_t stream0;
    cudaStreamCreate(&stream0);

    float* host_a;
    cudaMallocHost(&host_a, num_floats * sizeof(float));

    float* device_a;
    cudaMalloc(&device_a, num_floats * sizeof(float));

    auto start = std::chrono::high_resolution_clock::now();

    cudaMemcpyAsync(device_a, host_a, num_floats * sizeof(float), cudaMemcpyHostToDevice, stream0);
    cudaStreamSynchronize(stream0);

    auto end = std::chrono::high_resolution_clock::now();
    printf("Time: %f\n", std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / 1000.0);

    cudaFreeHost(host_a);
    cudaFree(device_a);
    cudaStreamDestroy(stream0);

    return 0;
}
```
With output:
```
Time: 0.401000
```

**Comment by lengstrom** [comment]
Going from CPU to GPU with Jax is similarly slow. However, unless I am doing something wrong (likely), going from numpy to GPU directly (instead of Jax CPU to GPU) increases throughput by 10x.. strange?

```python
import time
import jax
import jax.numpy as jnp
import numpy as np

cpu_dev = jax.devices('cpu')[0]
gpu_dev = jax.devices('gpu')[0]

GBs = 10
def big_tensor():
    # prev:
    # return jnp.ones((GBs * 1024**3 // 4,), dtype=jnp.float32,
    #                 device=jax.devices('gpu')[0])
    return np.ones((GBs * 1024**3 // 4,), dtype=np.float32)

def test_transfer(state):
    jax.block_until_ready(state)
    s = time.time()

    # prev:
    # out = jax.device_put(state, cpu_dev)
    out = jax.device_put(state, gpu_dev)

    jax.block_until_ready(out)
    dur = time.time() - s
    print(f"Time taken: {dur}; gbps: {GBs/dur}")

for i in range(5):
    test_transfer(big_tensor())
```

**Comment by granular-storage** [comment]
You"'re aware device_put is slower than it should be and looking at speeding it up soon" do you have a work in progress design or any progress on this? Thanks.
